internal/users/users.go

```diff
@@ -0,0 +1,176 @@
+// Package users support all common action on the system for user handling.
+package users
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"log/slog"
+	"os"
+	"os/exec"
+	"slices"
+	"strings"
+
+	"github.com/ubuntu/decorate"
+)
+
+// UserInfo is the user information returned by the broker. We use that to build our own buckets content.
+type UserInfo struct {
+	Name  string
+	UID   int
+	Gecos string
+	Dir   string
+	Shell string
+
+	Groups []GroupInfo
+}
+
+// GroupInfo is the group information returned by the broker. We use that to build our own buckets content.
+type GroupInfo struct {
+	Name string
+	GID  *int
+}
+
+type options struct {
+	groupPath  string
+	gpasswdCmd []string
+}
+
+var defaultOptions = options{
+	groupPath:  "/etc/group",
+	gpasswdCmd: []string{"gpasswd"},
+}
+
+// Option represents an optional function to override Cache default values.
+type Option func(*options)
+
+// UpdateLocalGroups synchronizes for the given user the local group list with the current group list from UserInfo.
+func (u *UserInfo) UpdateLocalGroups(args ...Option) (err error) {
+	defer decorate.OnError(&err, "could not update local groups for user %q", u.Name)
+
+	if u.Name == "" {
+		return errors.New("empty user name")
+	}
+
+	for _, g := range u.Groups {
+		if g.Name == "" && g.GID == nil {
+			return errors.New("empty group provided")
+		}
+	}
+
+	opts := defaultOptions
+	for _, arg := range args {
+		arg(&opts)
+	}
+
+	currentLocalGroups, err := existingLocalGroups(u.Name, opts.groupPath)
+	if err != nil {
+		return err
+	}
+
+	groupsToAdd, groupsToRemove := computeGroupOperation(u.Groups, currentLocalGroups)
+
+	for _, g := range groupsToRemove {
+		args := opts.gpasswdCmd[1:]
+		args = append(args, "--delete", u.Name, g)
+		if err := runGPasswd(opts.gpasswdCmd[0], args...); err != nil {
+			return err
+		}
+	}
+	for _, g := range groupsToAdd {
+		args := opts.gpasswdCmd[1:]
+		args = append(args, "--add", u.Name, g)
+		if err := runGPasswd(opts.gpasswdCmd[0], args...); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// existingLocalGroups returns which groups from groupPath the user is part of.
+func existingLocalGroups(user, groupPath string) (groups []string, err error) {
+	defer decorate.OnError(&err, "could not fetch existing local group")
+
+	f, err := os.Open(groupPath)
+	if err != nil {
+		return nil, err
+	}
+
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		t := strings.TrimSpace(scanner.Text())
+		if t == "" {
+			continue
+		}
+		elems := strings.Split(t, ":")
+		if len(elems) != 4 {
+			return nil, fmt.Errorf("unexpected number of elements in group file on line (should have 4 separators): %q", t)
+		}
```

Comment:
Can you add a comment with an example of a group entry? I know it is "well known" at this point, but I think it's good to have a visual representation in this kind of parsing (just so the hardcoding doesn't feel as bad)

e.g. ```// {group_name}:{x}:{group_id}:{user1, ..., usern}```